lib/stateful/animation.dart

```diff
@@ -0,0 +1,288 @@
+import 'dart:async';
+import 'dart:math' as math;
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/material.dart';
+
+class StatefulAnimation extends StatefulWidget {
+  StatefulAnimation({
+    Key key,
+  }): super(key: key);
+
+  @override
+  _StatefulAnimationState createState() => _StatefulAnimationState();
+}
+
+class _StatefulAnimationState extends State<StatefulAnimation> {
+  ValueNotifier<Color> _colorListenable;
+  Timer _timer;
+
+  @override
+  void initState() {
+    super.initState();
+    _colorListenable = ValueNotifier<Color>(Colors.green[100]);
+    int color = 100;
+    _timer = Timer.periodic(const Duration(milliseconds: 900), (Timer timer) {
+      color += 100;
+      if (color > 900)
+        color = 100;
+      _colorListenable.value = Colors.green[color];
+    });
+  }
+
+  @override
+  void dispose() {
+    _timer.cancel();
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Example(
+      restorationId: 'animation',
+      color: _colorListenable,
+      active: true,
+    );
+  }
+}
+
+class Example extends StatefulWidget {
+  Example({
+    Key key,
+    @required this.restorationId,
+    this.color,
+    this.active,
+  }): assert(restorationId != null),
+      assert(color != null),
+      assert(active != null),
+      super(key: key);
+
+  final String restorationId;
+
+  final ValueListenable<Color> color;
+
+  final bool active;
+
+  @override
+  _ExampleState createState() => _ExampleState();
+}
+
+class _ExampleState extends State<Example> with TickerProviderStateMixin, RestorationMixin {
+  RestorableDuration _duration = RestorableDuration(const Duration(seconds: 5));
+  AnimationController _animation;
+  Color _color;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.color.addListener(_colorChange);
+    _color = widget.color.value;
+  }
+
+  bool _firstTime = true;
+
+  @override
+  void restoreState(RestorationBucket oldBucket) {
+    registerForRestoration(_duration, 'duration');
+    if (_firstTime)
+      _animation = AnimationController(vsync: this);
+    _animation.duration = _duration.value;
+    if (_firstTime && widget.active)
+      _animation.repeat();
+    _firstTime = false;
+  }
+
+  @override
+  String get restorationId => widget.restorationId;
+
+  @override
+  void didUpdateWidget(Example oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    oldWidget.color.removeListener(_colorChange);
+    widget.color.addListener(_colorChange);
+    if (widget.active != oldWidget.active) {
+      if (widget.active) {
+        _animation.repeat();
+      } else {
+        _animation.stop();
+      }
+    }
+  }
+
+  double _height;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _height = _expensiveComputation(MediaQuery.of(context).size.height);
```

Comment:
nit but I never understood the point of having `didChangeDependencies`
It's very brittle as adding any new dependency on a different InheritedWidget will cause the performance of this code to drop

So in the end we'd want to do:

```dart
final sizeHeight = MediaQuery.of(context).size.height;
if (previousSizeHeight != sizeHeight) {
  previousSizeHeight = sizeHeight;
  _height = _expensiveComputation(sizeHeight);
}
```

But that's very tedious.